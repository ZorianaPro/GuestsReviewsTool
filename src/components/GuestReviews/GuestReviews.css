.GuestReviews {
  width: 100%;
  margin: 0 auto;
}

.GuestReviews .GuestReviews-Container {
  max-width: 90%;
  width: 1200px;
  margin: 10em auto 3em;
  position: relative;
  height: 100%
}

@media only screen and (max-width: 992px) {
  .GuestReviews .GuestReviews-Container {
    margin: 5em auto 3em;
  }
}

.GuestReviews .GuestReviews-Bg {
  height: 700px;
  position: absolute;
  width: 90%;
  top: 0;
  left: 50%;
  transform: translateX(-50%);
}

@media only screen and (max-width: 1400px) {
  .GuestReviews .GuestReviews-Bg {
    width: 100%;
  }
}

.GuestReviews .GuestReviews-Bg-Image {
  background-image: url('/background.png');
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  height: 100%;
  width: 100%;
  position: relative;
}

.GuestReviews .GuestReviews-Bg-Mask {
  height: 100%;
  width: 100%;
  background-color: rgba(0,0,0,0.3);
}

.GuestReviews .GuestReviews-Container-Main {
  width: 100%;
  display: flex;
  justify-content: flex-end;
}

@media only screen and (max-width: 992px) {
  .GuestReviews .GuestReviews-Container-Main {
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;
  }
}