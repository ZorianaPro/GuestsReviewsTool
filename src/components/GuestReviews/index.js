import React, { useEffect, useState, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getReviews } from '../../redux/reviews/actions';
import LoadMore from '../LoadMore';
import Filters from '../Filters';
import AirbnbSVG from '../AirbnbSVG';
import BookingcomSVG from '../BookingcomSVG';
import HoliduSVG from '../HoliduSVG';
import ThumbDownSVG from '../ThumbDownSVG';
import ThumbUpSVG from '../ThumbUpSVG';
import './GuestReviews.css';
const filters = require('../../services/filters');

const GuestReviews = () => {
  const [shownResults, setShownResults] = useState(0);
  const [totalResults, setTotalResults] = useState();
  const [shouldShowLoadButton, setShouldShowLoadButton] = useState(true);
  const [isLoading, setIsLoading] = useState();
  const [scoreFilter, setScoreFilter] = useState();
  const [channelFilter, setChannelFilter] = useState();

  const reviewsState = useSelector((state) => state.reviews);
  const dispatch = useDispatch();

  const fetchData = useCallback(
    (start = 0, limit = 10, filters) => {
      dispatch(getReviews({
        start: start,
        limit: limit,
        filters: filters
      }));
    }, [dispatch]);

  useEffect(() => {
    const filteringData = {
      channel: channelFilter || [],
      score: scoreFilter || ''
    };
    fetchData(0, 10, filteringData);
  }, [channelFilter, scoreFilter, fetchData]);

  useEffect(() => {
    setIsLoading(reviewsState.isLoading);
  }, [reviewsState.isLoading]);

  useEffect(() => {
    if (reviewsState.reviews && reviewsState.reviews.length) {
      setShownResults(reviewsState.reviews.length);
      setTotalResults(reviewsState.amount);
      if (reviewsState.amount === reviewsState.reviews.length) {
        setShouldShowLoadButton(false);
      }
    }
  }, [reviewsState.reviews]);

  useEffect(() => {
    if (reviewsState.reviews && reviewsState.reviews.length) {
      if (reviewsState.amount === reviewsState.reviews.length) {
        setShouldShowLoadButton(false);
      } else {
        setShouldShowLoadButton(true);
      }
    }
    setTotalResults(reviewsState.amount);
    if (reviewsState.amount === 0) {
      setShouldShowLoadButton(false);
    } else {
      setShouldShowLoadButton(true);
    }
  }, [reviewsState.amount]);

  const onCLickLoadMore =  useCallback(() => {
    const filteringData = {
      channel: channelFilter || [],
      score: scoreFilter || ''
    };
    if (shownResults !== totalResults) {
      fetchData(shownResults, 10, filteringData);
    }
  }, [shownResults, fetchData, totalResults, scoreFilter, channelFilter]);

  const onChangeScore = useCallback((event) => {
    event.preventDefault();
    const formData = new FormData(event.target);
    setScoreFilter(formData.get('score'));
  }, [setScoreFilter]);

  const onChangeChannel = useCallback((event) => {
    event.preventDefault();
    const formData = new FormData(event.target);
    setChannelFilter(formData.getAll('channel'));
  }, [setChannelFilter]);

  return (
    <div className="GuestReviews">
      <Filters
        filters={
          filters
        }
        onChangeScore={
          onChangeScore
        }
        onChangeChannel={
          onChangeChannel
        }
      />
      { totalResults }
      {
        reviewsState.reviews
        && reviewsState.reviews
          .map((review) =>
            <div className="Review">
              {
                review.headline
              }
            </div>
          )
      }
      {
        shouldShowLoadButton
        && <LoadMore
          onClick={
            onCLickLoadMore
          }
          isLoading={
            isLoading
          }/>
      }
      <AirbnbSVG/>
      <BookingcomSVG/>
      <HoliduSVG/>
      <ThumbDownSVG/>
      <ThumbUpSVG/>
    </div>
  );
};

export default GuestReviews;
