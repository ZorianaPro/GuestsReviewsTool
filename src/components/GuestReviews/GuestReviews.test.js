import React from 'react';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import GuestReviews from './';
import thunk from 'redux-thunk';

const mockStore = configureStore([thunk]);
const initialState = {
  reviews: {}
};

describe('GuestReviews', () => {
  let store;

  beforeAll(() => {
    store = mockStore(initialState);
  });

  it('renders without crashing', () => {
    expect(
      shallow.bind(
        shallow,
        <Provider store={ store }>
          <GuestReviews />
        </Provider>
      )
    )
      .not
      .toThrow();
  });

  it('renders as expected', () => {
    expect(
      mount(
        <Provider store={ store }>
          <GuestReviews />
        </Provider>
      )
    ).toMatchSnapshot();
  });
});
