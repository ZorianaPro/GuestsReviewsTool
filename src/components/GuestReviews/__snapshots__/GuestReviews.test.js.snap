// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuestReviews renders as expected 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <GuestReviews>
    <div
      className="GuestReviews"
    >
      <Filters
        filters={
          Array [
            Object {
              "name": "score",
              "values": Array [
                Object {
                  "value": "3",
                },
                Object {
                  "value": "3.5",
                },
                Object {
                  "value": "4",
                },
                Object {
                  "value": "4.5",
                },
              ],
            },
            Object {
              "name": "channel",
              "values": Array [
                Object {
                  "value": "AIRBNB",
                },
                Object {
                  "value": "HOLIDU",
                },
                Object {
                  "value": "BOOKINGCOM",
                },
              ],
            },
          ]
        }
        onChangeChannel={[Function]}
        onChangeScore={[Function]}
      >
        <div
          className="Filters"
        >
          <h4>
            Filters
          </h4>
          <form
            className="Filters-Category"
            name="score"
            onSubmit={[Function]}
          >
            <p>
              score
            </p>
            <div>
              <input
                id="score-0"
                name="score"
                onChange={[Function]}
                type="radio"
                value="3"
              />
              <label
                htmlFor="score-0"
              >
                3
                <span>
                   & up
                </span>
              </label>
            </div>
            <div>
              <input
                id="score-1"
                name="score"
                onChange={[Function]}
                type="radio"
                value="3.5"
              />
              <label
                htmlFor="score-1"
              >
                3.5
                <span>
                   & up
                </span>
              </label>
            </div>
            <div>
              <input
                id="score-2"
                name="score"
                onChange={[Function]}
                type="radio"
                value="4"
              />
              <label
                htmlFor="score-2"
              >
                4
                <span>
                   & up
                </span>
              </label>
            </div>
            <div>
              <input
                id="score-3"
                name="score"
                onChange={[Function]}
                type="radio"
                value="4.5"
              />
              <label
                htmlFor="score-3"
              >
                4.5
                <span>
                   & up
                </span>
              </label>
            </div>
            <button
              type="submit"
            />
          </form>
          <form
            className="Filters-Category"
            name="channel"
            onSubmit={[Function]}
          >
            <p>
              channel
            </p>
            <div>
              <input
                id="channel-0"
                name="channel"
                onChange={[Function]}
                type="checkbox"
                value="AIRBNB"
              />
              <label
                htmlFor="channel-0"
              >
                AIRBNB
              </label>
            </div>
            <div>
              <input
                id="channel-1"
                name="channel"
                onChange={[Function]}
                type="checkbox"
                value="HOLIDU"
              />
              <label
                htmlFor="channel-1"
              >
                HOLIDU
              </label>
            </div>
            <div>
              <input
                id="channel-2"
                name="channel"
                onChange={[Function]}
                type="checkbox"
                value="BOOKINGCOM"
              />
              <label
                htmlFor="channel-2"
              >
                BOOKINGCOM
              </label>
            </div>
            <button
              type="submit"
            />
          </form>
        </div>
      </Filters>
      <LoadMore
        onClick={[Function]}
      >
        <div
          className="LoadMore"
        >
          <div
            className="LoadMore-Button"
            onClick={[Function]}
          >
            <p>
              load more
            </p>
          </div>
        </div>
      </LoadMore>
      <AirbnbSVG>
        <svg
          className="AirbnbSVG"
        >
          <use
            xlinkHref="#AirbnbSymbol"
          />
        </svg>
      </AirbnbSVG>
      <BookingcomSVG>
        <svg
          className="BookingcomSVG"
        >
          <use
            xlinkHref="#BookingcomSymbol"
          />
        </svg>
      </BookingcomSVG>
      <HoliduSVG>
        <svg
          className="HoliduSVG"
        >
          <use
            xlinkHref="#HoliduSymbol"
          />
        </svg>
      </HoliduSVG>
      <ThumbDownSVG>
        <svg
          className="ThumbDownSVG"
        >
          <use
            xlinkHref="#ThumbDownSymbol"
          />
        </svg>
      </ThumbDownSVG>
      <ThumbUpSVG>
        <svg
          className="ThumbUpSVG"
        >
          <use
            xlinkHref="#ThumbUpSymbol"
          />
        </svg>
      </ThumbUpSVG>
    </div>
  </GuestReviews>
</Provider>
`;
